File: src/org/genyris/logic/OrFunction.java
package org.genyris.logic;

import org.genyris.core.Exp;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.GenyrisException;

public class OrFunction extends ApplicableFunction {

	public OrFunction(Interpreter interp) {
		super(interp);
	}

	public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment envForBindOperations)
			throws GenyrisException {
		if (arguments.length < 2)
			throw new GenyrisException("Too few arguments to or: " + arguments.length);

		for (int i = 0; i < arguments.length; i++) {
			if (arguments[i] == TRUE) {
				return TRUE;
			}
			else if (arguments[i] == NIL) {
				continue;
			}
			else {
				throw new GenyrisException("or expects t or nil, not: " + arguments[i]);				
			}
		}
		return NIL;

	}

}

File: src/org/genyris/logic/OrFunction.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.logic

import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.ApplicableFunction
import org.genyris.interp.Closure
import org.genyris.interp.Environment
import org.genyris.interp.Interpreter

class OrFunction(interp: Interpreter) : ApplicableFunction(interp, "or", false) {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>, envForBindOperations: Environment?): Exp? {
        if (arguments.size < 2) throw GenyrisException("Too few arguments to or: " + arguments.size)
        var result: Exp? = NIL
        for (i in arguments.indices) {
            result = arguments[i]!!.eval(envForBindOperations)
            if (result !== NIL) {
                return result
            } else {
                continue
            }
        }
        return NIL
    }
}
